packages/flutter/lib/src/widgets/framework.dart

```diff
@@ -3731,7 +3731,17 @@ class StatefulElement extends ComponentElement {
     assert(_state._debugLifecycleState == _StateLifecycle.created);
     try {
       _debugSetAllowIgnoredCallsToMarkNeedsBuild(true);
-      _state.initState();
+      final void _debugCheckForReturnedFuture = _state.initState();
+      assert(() {
+        if ((_debugCheckForReturnedFuture  as dynamic) is Future) {
```

Comment:
Double-space before `as`. Also, `as` is banned in Flutter :)

Reply:
Ahh this is really the crux of this solution.  For a function which is annotated `void foo() async`, you cannot verify that it has an async keyword without checking the runtime type of the returned value.  Statically, the function is still a `void Function()`;

Reply:
I've updated the code to match the implementation of the setState return check (which does indeed use as).

Reply:
I don't know what exact semantics Dart 2 is aiming at w.r.t `void`. If `void` is capable of carrying a value then I'd expect that either `is` check would be allowed without the `as`, or `somethingOfTypeVoid as` would become illegal. @leafpetersen likely knows.

Reply:
Basically, I'd expect that the language would prevent the user of a `void` function from observing its return type or value. Otherwise, changing from `async` to non-`async` becomes a breaking API change.

Reply:
Just as an aside, if this won't work in Dart 2 then the setState return value check will need to be reworked as well.

Reply:
Didn't see your comment about `setState`. Yeah, if we do this already elsewhere, then it makes sense to add a similar check here. That said, while there's almost never a good reason to do async work inside a `setState` callback, there are plenty of use-cases to do it as part of `initState`.

Reply:
You could consider typing this as `FutureOr<void> _debugCheckForReturnedFuture`.

It breaks too much existing code to stop void from carrying a value statically.  Everyone would like to but... :(

We plan to allow (`returnsVoid() as T`) but not (`returnsVoid is T`).  It's not something you should be doing, so we'd prefer you have to have to go out of your way to do it.

@lrhn  a slightly interesting observation that your proposal to dynamically drop the return value of void typed things would break this check.

Reply:
FYI https://github.com/flutter/flutter/pull/15183

Also a similar check is currently being used to verify that setState (a void typedef) doesn't return a future